# Strategy Name: Monthly Dollar-Cost Averaging in SPY ETF

### Strategy Overview: 
The Monthly Dollar-Cost Averaging (DCA) strategy is designed to systematically invest a fixed amount of capital into the SPDR S&P 500 ETF Trust (SPY) on a monthly basis. This approach aims to mitigate the impact of market volatility and reduce the risk associated with timing the market.

### Investment Objective: 
The primary objective of this strategy is to accumulate a significant position in the SPY ETF over time while taking advantage of potential price fluctuations. By consistently investing a fixed amount each month, the strategy seeks to benefit from averaging down during market dips and capturing upside during bullish trends.

### Time Period:
* Start Date: October 1, 2019
* End Date: October 1, 2024
* Duration: 5 years
* Investment Amount:

* Fixed Monthly Investment: $500

### Execution Methodology:

* Investment Frequency: Investments will be made on the first trading day of each month.
* Asset: SPDR S&P 500 ETF Trust (SPY), which aims to track the performance of the S&P 500 Index.
* Price Point for Investment: The closing price of SPY on the first trading day of each month.

### Performance Metrics: 
The performance of the strategy will be evaluated based on:

* Total Investment Amount: Cumulative capital deployed over the investment period.
* Current Portfolio Value: Value of the accumulated SPY shares at the end of the investment period.
* Total Profit: Difference between the current portfolio value and the total investment amount.

### Rationale: 
This strategy leverages the principle of dollar-cost averaging, which is particularly effective in volatile markets. By investing consistently over time, the strategy aims to reduce the emotional impact of market fluctuations and promote disciplined investing behavior. The SPY ETF is selected for its broad market exposure, liquidity, and historical performance.

### Risk Considerations:

* Market risk: The strategy is subject to the inherent volatility of the equity markets.
* Opportunity cost: Fixed monthly investments may result in lower returns compared to a lump-sum investment in a rising market.
* Transaction costs: While investing monthly may incur trading fees, these should be minimized through appropriate brokerage choices.

### Conclusion: 

The Monthly Dollar-Cost Averaging strategy provides a disciplined approach to investing in the SPY ETF, allowing investors to build wealth gradually while managing risk. This strategy is particularly suitable for long-term investors seeking to participate in the growth of the U.S. equity market without the stress of market timing.

## Step 1: Fetch Historical Data for SPY
First, let’s fetch the historical data for SPY over the last 5 years.

In [1]:
import pandas as pd
import yfinance as yf

# Fetch SPY data for the last 5 years
symbol = 'SPY'
start_date = '2019-10-01'  # 5 years ago
end_date = '2024-10-01'

data = yf.download(symbol, start=start_date, end=end_date)

data

[*********************100%%**********************]  1 of 1 completed


Open        High         Low       Close   Adj Close  \
Date                                                                     
2019-10-01  297.739990  298.459991  293.000000  293.239990  271.450409   
2019-10-02  291.500000  291.510010  286.640015  288.059998  266.655304   
2019-10-03  287.809998  290.450012  284.820007  290.420013  268.839996   
2019-10-04  291.140015  294.630005  291.079987  294.350006  272.477966   
2019-10-07  293.470001  295.260010  292.769989  293.079987  271.302216   
...                ...         ...         ...         ...         ...   
2024-09-24  570.479980  571.359985  567.599976  571.299988  571.299988   
2024-09-25  571.140015  571.890015  568.909973  570.039978  570.039978   
2024-09-26  574.380005  574.710022  569.900024  572.299988  572.299988   
2024-09-27  573.390015  574.219971  570.419983  571.469971  571.469971   
2024-09-30  570.419983  574.380005  568.080017  573.760010  573.760010   

               Volume  
Date                   
2019-10-01   88242400  
2019-10-02  122539500  
2019-10-03   85278800  
2019-10-04   65091200  
2019-10-07   59610500  
...               ...  
2024-09-24   46805700  
2024-09-25   38428600  
2024-09-26   48336000  
2024-09-27   42100900  
2024-09-30   63557400  

[1258 rows x 6 columns]

## Step 2: Resample Data to Get First Trading Day of Each Month
We'll get the first trading day of each month and the closing price on that day.

In [2]:
# Get the first trading day of each month
monthly_data = data.resample('MS').first()  # 'BM' means Business Month end frequency
monthly_data

Open        High         Low       Close   Adj Close  \
Date                                                                     
2019-10-01  297.739990  298.459991  293.000000  293.239990  271.450409   
2019-11-01  304.920013  306.190002  304.739990  306.140015  283.391846   
2019-12-01  314.589996  314.660004  311.170013  311.640015  288.483215   
2020-01-01  323.540009  324.890015  322.529999  324.869995  302.208649   
2020-02-01  323.350006  326.160004  323.220001  324.119995  301.510986   
2020-03-01  298.209991  309.160004  294.459991  309.089996  287.529388   
2020-04-01  247.979996  257.660004  243.899994  246.149994  230.326233   
2020-05-01  285.309998  290.660004  281.519989  282.790009  264.610809   
2020-06-01  303.619995  306.209991  303.059998  305.549988  285.907654   
2020-07-01  309.570007  311.890015  309.070007  310.519989  291.836761   
2020-08-01  328.320007  329.619995  327.730011  328.790009  309.007599   
2020-09-01  350.209991  352.709991  349.239990  352.600006  331.385010   
2020-10-01  337.690002  338.739990  335.010010  337.040009  318.029053   
2020-11-01  330.200012  332.359985  327.239990  330.200012  311.575012   
2020-12-01  365.570007  367.679993  364.929993  366.019989  345.374512   
2021-01-01  375.309998  375.450012  364.820007  368.790009  349.471649   
2021-02-01  373.720001  377.339996  370.380005  376.230011  356.521973   
2021-03-01  385.589996  390.920013  380.570007  389.579987  369.172607   
2021-04-01  398.399994  400.670013  398.179993  400.609985  380.868134   
2021-05-01  419.429993  419.839996  417.670013  418.200012  397.591370   
2021-06-01  422.570007  422.720001  419.200012  419.670013  398.988953   
2021-07-01  428.869995  430.600006  428.799988  430.429993  410.557495   
2021-08-01  440.339996  440.929993  437.209991  437.589996  417.386902   
2021-09-01  452.559998  453.109985  451.549988  451.799988  430.940826   
2021-10-01  430.980011  436.029999  427.230011  434.239990  415.518494   
2021-11-01  460.299988  460.700012  458.200012  460.040009  440.206146   
2021-12-01  461.640015  464.670013  450.290009  450.500000  431.077484   
2022-01-01  476.299988  477.850006  473.850006  477.709991  458.720245   
2022-02-01  450.679993  453.630005  446.940002  452.950012  434.944550   
2022-03-01  435.040009  437.170013  427.109985  429.980011  412.887695   
2022-04-01  453.309998  453.459991  449.140015  452.920013  436.266876   
2022-05-01  412.070007  415.920013  405.019989  414.480011  399.240265   
2022-06-01  415.170013  416.239990  406.929993  409.589996  394.530029   
2022-07-01  376.559998  381.700012  373.799988  381.239990  368.808746   
2022-08-01  409.149994  413.410004  408.399994  410.769989  397.375824   
2022-09-01  392.890015  396.779999  390.040009  396.420013  383.493744   
2022-10-01  361.079987  368.549988  359.209991  366.609985  356.112640   
2022-11-01  390.140015  390.390015  383.290009  384.519989  373.509796   
2022-12-01  408.769989  410.000000  404.750000  407.380005  395.715271   
2023-01-01  384.369995  386.429993  377.829987  380.820007  371.614410   
2023-02-01  405.209991  413.670013  402.350006  410.799988  400.869720   
2023-03-01  395.410004  396.690002  393.380005  394.739990  385.197906   
2023-04-01  408.850006  411.369995  408.440002  410.950012  402.546600   
2023-05-01  415.470001  417.619995  415.269989  415.510010  407.013336   
2023-06-01  418.089996  422.920013  416.790009  421.820007  413.194275   
2023-07-01  442.920013  444.079987  442.630005  443.790009  436.329803   
2023-08-01  456.269989  457.250000  455.489990  456.480011  448.806488   
2023-09-01  453.170013  453.670013  449.679993  451.190002  443.605438   
2023-10-01  426.619995  428.600006  424.459991  427.309998  421.608765   
2023-11-01  419.200012  423.500000  418.649994  422.660004  417.020844   
2023-12-01  455.769989  459.649994  455.160004  459.100006  452.974640   
2024-01-01  472.160004  473.670013  470.489990  472.649994  468.234589   
2024-02-01  484.630005

## Step 3: Simulate the Investment
Now we’ll simulate investing $500 at the close price of SPY on the first trading day of each month.

In [3]:
# Investment parameters
investment_per_month = 500
total_investment = 0
shares_purchased = []

# Iterate through the monthly data to simulate the investment
for date, row in monthly_data.iterrows():
    close_price = row['Close']
    shares = investment_per_month / close_price  # Calculate shares purchased
    shares_purchased.append(shares)
    total_investment += investment_per_month

# Calculate total shares bought
total_shares = sum(shares_purchased)

# Calculate the current value of the investment
current_value = total_shares * data['Close'][-1]  # Use the last available close price

current_value

C:\Users\roylq\AppData\Local\Temp\ipykernel_16512\2250151110.py:17: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  current_value = total_shares * data['Close'][-1]  # Use the last available close price


43465.6854871773

## Step 4: Calculate Results
Now let’s calculate how much you would have now and how much profit you made.

In [4]:
profit = current_value - total_investment

print(f'Total Investment: ${total_investment}')
print(f'Current Value: ${current_value:.2f}')
print(f'Profit: ${profit:.2f}')


Total Investment: $30000
Current Value: $43465.69
Profit: $13465.69


## Full code packages: 

In [5]:
import pandas as pd
import yfinance as yf

# Fetch SPY data for the last 5 years
symbol = 'SPY'
start_date = '2019-10-01'  # 5 years ago
end_date = '2024-10-01'

data = yf.download(symbol, start=start_date, end=end_date)

# Get the first trading day of each month
monthly_data = data.resample('MS').first()

# Investment parameters
investment_per_month = 500
total_investment = 0
shares_purchased = []

# Simulate the investment
for date, row in monthly_data.iterrows():
    close_price = row['Close']
    shares = investment_per_month / close_price
    shares_purchased.append(shares)
    total_investment += investment_per_month

# Calculate total shares bought and current value
total_shares = sum(shares_purchased)
current_value = total_shares * data['Close'][-1]  # Use the last available close price

# Results
profit = current_value - total_investment

print(f'Total Investment: ${total_investment}')
print(f'Current Value: ${current_value:.2f}')
print(f'Profit: ${profit:.2f}')


[*********************100%%**********************]  1 of 1 completed

Total Investment: $30000
Current Value: $43465.69
Profit: $13465.69



C:\Users\roylq\AppData\Local\Temp\ipykernel_16512\2576967782.py:28: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  current_value = total_shares * data['Close'][-1]  # Use the last available close price


## Robust analysis: 

In real-world investing, particularly with ETFs like SPY, an investor can only purchase whole shares. If a monthly investment does not cover the price of a full share, the investor would only buy as many whole shares as possible, and any remaining funds would carry over to the next month's investment.

### Strategy Adjustment Description
* Adjustment to Investment Execution: To accurately reflect the behavior of an individual investor in the dollar-cost averaging strategy, the following adjustments will be made:

* Whole Shares Only: Each monthly investment will be calculated to purchase only whole shares of SPY.
Carryover Funds: Any unused funds from a month's investment (if the fixed amount is insufficient to purchase at least one full share) will be added to the subsequent month's investment.

In [6]:
import pandas as pd
import yfinance as yf

# Fetch SPY data for the last 5 years
symbol = 'SPY' # This code could be replace by other ETF, Fund, Stock
start_date = '2019-10-01'  # 5 years ago
end_date = '2024-10-01'

data = yf.download(symbol, start=start_date, end=end_date)

# Get the first trading day of each month
monthly_data = data.resample('MS').first()  # 'MS' for Month Start

# Investment parameters------------------------------------ This part being edit -----------------------------------
investment_per_month = 500
total_investment = 0
shares_purchased = []
carryover = 0  # Initialize carryover

# Simulate the investment
for date, row in monthly_data.iterrows():
    close_price = row['Close']
    
    # Total investment for the month, including any carryover
    total_monthly_investment = investment_per_month + carryover
    
    # Calculate shares purchased (whole shares only)
    shares = total_monthly_investment // close_price  # Integer division for whole shares
    shares_purchased.append(shares)
    
    # Calculate the amount used for shares and the carryover
    amount_used = shares * close_price
    carryover = total_monthly_investment - amount_used  # Remaining funds to carry over
    
    # Update total investment
    total_investment += investment_per_month

# Calculate total shares bought and current value
total_shares = sum(shares_purchased)
current_value = total_shares * data['Close'][-1]  # Use the last available close price
# -----------------------------------------------------------------------------------------------------------
# Results
profit = current_value - total_investment

print(f'Total Investment: ${total_investment}')
print(f'Current Value: ${current_value:.2f}')
print(f'Profit: ${profit:.2f}')



[*********************100%%**********************]  1 of 1 completed

Total Investment: $30000
Current Value: $43032.00
Profit: $13032.00



C:\Users\roylq\AppData\Local\Temp\ipykernel_16512\2465601732.py:40: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  current_value = total_shares * data['Close'][-1]  # Use the last available close price


## Stategy_1_Monthly Dollar-Cost Averaging in SPY ETF

#### Idealy: (if invest could buy the stock not in whole unit) 

* Total Investment: $30000

* Current Value: $43465.69

* Profit: $13465.69

#### Reality:(Buy the whold unit every month and put the reminder for next month) 

* Total Investment: $30000

* Current Value: $43032.00

* Profit: $13032.00